src/datasets/builder.py

```diff
@@ -654,19 +664,20 @@ def download_and_prepare(
             @contextlib.contextmanager
             def incomplete_dir(dirname):
                 """Create temporary dir for dirname and rename on exit."""
-                if is_remote_url(dirname):
-                    yield dirname
-                else:
-                    tmp_dir = dirname + ".incomplete"
-                    os.makedirs(tmp_dir, exist_ok=True)
-                    try:
-                        yield tmp_dir
-                        if os.path.isdir(dirname):
-                            shutil.rmtree(dirname)
-                        os.rename(tmp_dir, dirname)
-                    finally:
-                        if os.path.exists(tmp_dir):
-                            shutil.rmtree(tmp_dir)
+                tmp_dir = dirname + ".incomplete"
+                self._fs.makedirs(tmp_dir, exist_ok=True)
+                try:
+                    yield tmp_dir
+                    if self._fs.isdir(dirname):
+                        self._fs.rm(dirname, recursive=True)
+                    if is_local:
+                        # LocalFileSystem.mv does copy + rm, it is more efficient to simply rename a local directory
+                        os.rename(self._fs._strip_protocol(tmp_dir), self._fs._strip_protocol(dirname))
+                    else:
+                        self._fs.mv(tmp_dir, dirname, recursive=True)
```

Comment:
`mv` in `s3fs` and `hffs` (I expect these two filesystems + `gcsfs` to be the most used ones with this feature) reuploads the files, which is super expensive, so it would be great if we could replace this with some cheaper logic.

Reply:
I don't see an easy way to rename the files without reuploading in the general case, I guess we can just not use the `incomplete_dir` for cloud storage, what do you think ?

Reply:
Yes, I think this is the only option we have. (We just need to make sure we delete all the uploaded artifacts on error.)

Reply:
`mv` calls `cp_file` and `cp_file` is also optimized in `s3fs` (it calls [this API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html)), so I think it's also ok to keep this as-is for now and instead we should optimize `cp_file` in `hffs`

Reply:
oh that's good to know for s3fs, I'm down to keeping it as is then for simplicity, as we have a solution for hffs